Lib/posixpath.py

```diff
@@ -450,15 +453,17 @@ def realpath(filename, *, strict=False):
         else:
             newpath = path + sep + name
         try:
-            st = os.lstat(newpath)
-            if not stat.S_ISLNK(st.st_mode):
-                path = newpath
-                continue
+            st_mode = os.lstat(newpath).st_mode
         except OSError:
             if strict:
                 raise
             path = newpath
             continue
+        if not stat.S_ISLNK(st_mode):
+            if strict and part_count and not stat.S_ISDIR(st_mode):
+                raise NotADirectoryError(errno.ENOTDIR, "Not a directory", newpath)
```

Comment:
We could use `os.scandir(newpath).close()` to get a native error.

Reply:
What's the advantage of a native error, if you don't mind me asking?

Reply:
They may be more consistent with other OS errors. It is not always "Not a directory". But it does not matter much.